Xamarin.Forms.Core/BindableObject.cs

```diff
@@ -19,6 +19,8 @@ public abstract class BindableObject : INotifyPropertyChanged, IDynamicResourceH
 		bool _applying;
 		object _inheritedContext;
 
+		public Guid WindowId { get; set; }
```

Comment:
It would be ideal if we could hide this from users somehow

Reply:
We could hide it by creating a kind of Dispatcher object on the VisualElement, which would be something very similar to what we have in a native UWP or WPF application. We did a quick example based on this pattern: https://github.com/edguer/Xamarin.Forms/blob/multiple-window-uwp-2/Xamarin.Forms.Core/VisualElement.cs

In the example, the WindowId is still there, although as read-only for the users. We could spend more time on implementing these changes in a way we could eliminate the WindowId from our BindableObjects, but this would involve deeper changes on our current structure. The BeingInvokeOnMainThread and IsInvokeRequired, for example, should not be static anymore, reason being they would be located in the BindableObject Dispatcher object. Again, I'm thinking in a structure similar to the one in UWP.

Let me know your thoughts about it, based on your feedback we could initiate these changes as soon as possible.

Reply:
I have a proposal for this.

If we add a WindowId property to every BindableObject, that property will go completely unused on some of the platforms (at least for now). Even on platforms with multiple windows, it will be unused on _most_ BindableObjects; we only occasionally need it.

Instead, what if we provided a class which could globally provide a lookup for the relationship between an Element and Dispatcher? Something like

```
static class DispatcherRegistry
{
	IDispatcher GetDispatcher(Element element){...}
}
```

For the currently-single-window platforms, the implementation simply returns the same dispatcher it always did. For multi-window platforms, the implementation could use something like `ConditionalWeakTable<Element, Dispatcher>` to do the lookup.

Instead of setting a WindowId for a top-level element, we would add an entry into this table. When looking up a specific Element, we'd check for it in the table; if it's not in the table, we'd walk up the hierarchy to find the root (via Parent->Parent->...), which _will_ be in the table. Once we've determined the Dispatcher for an Element, we add a specific Entry for it in the table (i.e., cache for future lookups). This scheme avoids polluting Forms classes with a new property which most of them never need. 

The dispatcher we're getting back could implement

```
interface IDispatcher 
{
	void BeginInvokeOnMainThread(Action action);
	bool IsInvokeRequired();	
}
```

which for some of the platforms would just fall back to the existing PlatformServices methods. For the multi-window ones, the implementation can wrap the native Dispatcher.

We can add a convenience extension method for Element for each of these, something like:

```
public static class ElementExtensions
{
	public static Dispatcher Dispatcher(this Element element)
	{
		return DispatcherRegistry.GetDispatcher(element);
	}
}
```

So that it's clear and simple for people to use (similar to what @edguer suggested, but without adding the property directly to the Element class):

```
element.Dispatcher().BeginInvokeOnMainThread(() => DoStuff());
```

(Are extension properties in C# a thing yet? Maybe we can even get rid of that `()` after `Dispatcher`.)

Reply:
The proposal seems very good to me.

Extension properties are not a thing yet, but I don't see why we need to create an extension method for it.
Instead, I would make element.Dispatcher() a read-only property with our logic (IDispatcher Dispatcher { get {  return DispatcherRegistry.GetDispatcher(this); } }) and, if the platform doesn't need additional dispatcher or doesn't support multi-window, we will always return the same dispatcher. 

Of course this is the perspective of an outsider, so just let me know if this makes sense.

Reply:
I created a new branch with these changes: https://github.com/edguer/Xamarin.Forms/tree/multiple-window-uwp-refactored

In summary, we now have a Dispatcher property for each one of the VisualElements, and it only loads the Dispatcher in memory if it is called (afterwards it gets stored in a private field). That way, we didn't create any lookup table to store the dispatchers.
Also, we are only storing the WindowId for Application and Page, that way we avoid unnecessary properties in every BindableObject.

Let me know your thoughts so I can update the pull request.